lib/spack/spack/util/package_hash.py

```diff
@@ -74,82 +83,206 @@ def visit_ClassDef(self, node):  # noqa
+        if not any(r is func for r in resolutions):
+            # multimethod did not resolve to this function; remove it
+            return None
+
+        # if we get here, this function is a possible resolution for a multi-method.
+        # it might be the only one, or there might be several that have to be evaluated
+        # dynamcially.  Either way, we include the function.
 
+        # strip the when decorators (preserve the rest)
+        func.decorator_list = [
+            dec for dec in func.decorator_list
+            if not (isinstance(dec, ast.Call) and dec.func.id == 'when')
+        ]
+        return func
 
-def package_content(spec):
-    return ast.dump(package_ast(spec))
 
+def canonical_source(spec, filter_multimethods=True, source=None):
+    """Get canonical source for a spec's package.py by unparsing its AST.
 
-def package_hash(spec, content=None):
-    if content is None:
-        content = package_content(spec)
-    return hashlib.sha256(content.encode('utf-8')).digest().lower()
+    Arguments:
+        filter_multimethods (bool): By default, filter multimethods out of the
+            AST if they are known statically to be unused. Supply False to disable.
+        source (str): Optionally provide a string to read python code from.
+    """
+    return unparse(
+        package_ast(spec, filter_multimethods, source=source),
+        py_ver_consistent=True,
+    )
 
 
-def package_ast(spec):
+def package_hash(spec, source=None):
+    """Get a hash of a package's canonical source code.
+
+    This function is used to determine whether a spec needs a rebuild when a
+    package's source code changes.
+
+    Arguments:
+        source (str): Optionally provide a string to read python code from.
+
+    """
+    source = canonical_source(spec, filter_multimethods=True, source=source)
+    return spack.util.hash.b32_hash(source)
+
+
+def package_ast(spec, filter_multimethods=True, source=None):
+    """Get the AST for the ``package.py`` file corresponding to ``spec``.
+
+    Arguments:
+        filter_multimethods (bool): By default, filter multimethods out of the
+            AST if they are known statically to be unused. Supply False to disable.
+        source (str): Optionally provide a string to read python code from.
+    """
     spec = spack.spec.Spec(spec)
 
-    filename = spack.repo.path.filename_for_package_name(spec.name)
-    with open(filename) as f:
-        text = f.read()
-        root = ast.parse(text)
+    if source is None:
+        filename = spack.repo.path.filename_for_package_name(spec.name)
+        with open(filename) as f:
+            source = f.read()
 
+    root = ast.parse(source)
     root = RemoveDocstrings().visit(root)
-
     RemoveDirectives(spec).visit(root)
 
-    fmm = TagMultiMethods(spec)
-    fmm.visit(root)
+    if filter_multimethods:
+        tagger = TagMultiMethods(spec)
+        tagger.visit(root)
+        root = ResolveMultiMethods(tagger.methods).visit(root)
```

Comment:
Can we clean up the calling convention for these? It looks like `RemoveDirectives.visit` and `TagMultiMethods.visit` operate in place, but `RemoveDocstrings.visit` and `ResolveMultiMethods.visit` return a copy

Reply:
Yep this is a good idea.

Reply:
The tagger is just a visitor and doesn't modify the tree, so I didn't tweak that one, but the other two should be assignments because they are `NodeTransformer` subclasses.  I added some comments to make this clearer